### PR TITLE
Add MPL_mkstemp and cleanup leftovers

### PR DESCRIPTION
**Pull request by raffenet**
## Pull Request Description

Provide a portable mkstemp function in MPL and remove various leftover configure checks, declarations, and fallback code in MPICH.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

File: src/mpl/src/misc/Makefile.mk [hunk 1]
```diff
@@ -4,4 +4,5 @@
 #     See COPYRIGHT in top-level directory.
 #
 
-lib@MPLLIBNAME@_la_SOURCES += src/misc/mpl_misc.c
+lib@MPLLIBNAME@_la_SOURCES += src/misc/mpl_misc.c \
+                              src/misc/mpl_mkstemp.c
```
Comment (hzhou) [inline src/mpl/src/misc/Makefile.mk:8]: Did you forget to add `src/misc/mpl_mkstemp.c`?

Comment (raffenet) [reply src/mpl/src/misc/Makefile.mk:8]: indeed i did. added now.

**Comment by raffenet** [comment]
test:mpich/ch3/most

File: src/mpl/src/misc/mpl_mkstemp.c [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+#ifndef MPL_HAVE_MKSTEMP
+
+#warning The mkstemp function is not provided in the standard library.
+#warning Using a possibly insecure replacement function.
+
+#if !defined (HAVE_RAND) || !HAVE_RAND
+static notrand = 0;
+#define srand(x) notrand = x
+#define rand() (++notrand)
+#endif
```
Comment (hzhou) [inline src/mpl/src/misc/mpl_mkstemp.c:26]: `rand` is in standard C. How about remove the fallback and let's assume we always will have them?

File: src/mpl/src/misc/mpl_mkstemp.c [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+#include <fcntl.h>
+#include <unistd.h>
+
+/* here to prevent "has no symbols" warnings from ranlib on OS X */
+static int dummy ATTRIBUTE((unused)) MPL_USED = 0;
+
+#ifndef MPL_HAVE_MKSTEMP
+
+#warning The mkstemp function is not provided in the standard library.
+#warning Using a possibly insecure replacement function.
```
[the remark this answers is not in the document]
Comment (raffenet) [reply src/mpl/src/misc/mpl_mkstemp.c:20]: Is that true even though this is a .c file? Moving to configure is probably better either way. I'll add a cleanup patch.

[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpl/src/misc/mpl_mkstemp.c:20]: I was mistaken. I thought this is one of the MPL headers. In that case, I don't care either way. Up to you.

**Review by hzhou — APPROVED** [review]
LGTM

**Comment by raffenet** [comment]
squashed fixup. run a sanity check
test:mpich/ch3/tcp

**Pull request merged**